packages/components/src/custom-select-control/test/index.js

```diff
@@ -87,4 +88,127 @@ describe( 'CustomSelectControl', () => {
+
+		render(
+			<CustomSelectControl
+				onChange={ ( { selectedItem } ) => mock( selectedItem ) }
+				options={ options }
+				__nextUnconstrainedWidth
+			/>
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.click( currentValue );
+
+		const newValue = screen.getByRole( 'option', {
+			name: /poppy/i,
+		} );
+
+		await user.click( newValue );
+
+		expect( currentValue ).toHaveTextContent( /poppy/i );
+		expect( mock ).toHaveBeenCalled();
+	} );
+
+	it( 'Should be able to change selection using keyboard', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		expect( currentValue ).toHaveFocus();
+
+		await user.keyboard( '{enter}' );
+		await user.keyboard( '{arrowdown}' );
+		await user.keyboard( '{enter}' );
+
+		expect( currentValue ).toHaveTextContent( /crimson clover/i );
+	} );
+
+	it( 'Should keep current selection if dropdown is closed without changing selection', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		await user.keyboard( '{enter}' );
+		expect( screen.getByRole( 'listbox' ) ).toBeVisible();
+		expect( screen.getByRole( 'listbox' ) ).toHaveAttribute(
+			'aria-hidden',
+			'false'
+		);
+		await user.keyboard( '{escape}' );
+		expect( screen.queryByRole( 'listbox' ) ).not.toBeInTheDocument();
+
+		expect( currentValue ).toHaveTextContent( /violets/i );
+	} );
+
+	it( 'Should be able to type characters to select matching options', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		await user.keyboard( '{enter}' );
+		await user.keyboard( '{p}' );
+		await user.keyboard( '{enter}' );
+
+		expect( currentValue ).toHaveTextContent( /poppy/i );
+	} );
+
+	it( 'Can change selection with a focused input and closed dropdown if typed characters match an option', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		await user.keyboard( '{c}' );
+		await user.keyboard( '{enter}' );
+
+		expect( currentValue ).toHaveTextContent( /crimson clover/i );
+	} );
+
+	it( 'Current selection has attribute: aria-selected="true"', async () => {
```

Comment:
One way to make this test more valuable is to:
- make sure that the options that are not supposed to be selected have the `aria-selected="false"` attribute
- add some user interaction to select another option, and make sure that the `aria-selected` attributes update accordingly (this may fail if what I noticed in https://github.com/WordPress/gutenberg/pull/56575/files#r1411951559 is true)